text/0000-ember-language-tooling-roadmap.md

```diff
@@ -0,0 +1,169 @@
+
+In particular, the GlimmerX way of writing templates with Template Imports is:
+
+1. Write your components in `.ts` files (which by default triggers vanilla TS language server in most editors)
+2. Use `hbs`-tagged template literals to defining templates. Any variables/references within the template close over any import statements and variables declared in the file.
+
+The problem with this approach is that:
+
+1. When Vanilla TS runs, it'll produce errors/warnings such as "unused reference" for all the variables/imports that are only referenced within the `hbs` template, because it has no way at this point of knowing about GlimmerX's special Template Imports syntax/behavior.
+2. TS Plugins can only augment the set of errors within a file, they can't remove errors already added by default TS (or other plugins)
+3. Volar is moving away from "takeover mode" language servers to TS Plugins (which has many benefits described below), and there's no way to bring GlimmerX along for the ride
+
+Ember avoids these issues by way of a novel `.gts` / `.gjs` file extension which does not trigger vanilla TS tooling. This will allow us to implement a TS Plugin to process the file "from scratch" so that there are no default TS errors / warnings to remove.
+
+In conclusion, most people don't even know what GlimmerX is, it's not under active development, it doesn't play nicely with our plans to use TS Plugins. Therefore, we should remove it from Glint before proceeding further with the proposals in this RFC.
+
+### Proposal 2: Convert Glint from yarn to pnpm
+
+This is a small chore and will unify Glint with most modern repos in the Ember community as well as Volar's repos.
+
+### Proposal 3: Convert Glint to be backed by Volar 2.0, a framework for building language tooling that's used by Vue, MDX, Astro, and more.
+
+This is a major milestone and where the bulk of the work lies.
+
+Some of this work has been done in my [spike](https://github.com/machty/glint/tree/volar-spike), but most of that work is focused on .gts templates (no attempt has been made to get class-backed components + separate .hbs templates working -- see below for a discussion as to whether we should pursue that).
+
+For this milestone, we will simply attempt to achieve feature parity with existing Glint (but likely we will get a lot of commands/refactors/other functionality for free just by moving to Volar). In particular, this milestone continue to run Glint as a Language Server (rather than TS Plugin).
+
+### Proposal 4: Migrate Glint to provide editor completions, type-checking, etc, using a TS Plugin
+
+Vue Language Tooling has deprecated their Language-Server-backed language tooling in favor of a TS Plugin -driven tooling (while continuing to maintain vue-tsc for actually compilation of files).
```

Comment:
Would Glint continue to fulfil the same role that vue-tsc does for Vue? Does glint even get involved with compilation?

Reply:
For libraries, it's been needed to generate declarations, but support for gts is rough, and has many issues. I've been recommending post processing the declarations via 
 https://github.com/NullVoxPopuli/fix-bad-declaration-output

(Note tho that one of those issues is a TS issue rather than Glint, in that they don't believe in the concept of peer provided types)